include/roboteam_utils/Line.h

```diff
@@ -14,176 +14,196 @@ class LineSegment;
+     */
+    [[nodiscard]] Vector2 project(const Vector2 &point) const;
+
+    /**
+     * @brief Gets the intersection of the lines
+     * See https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection for help
+     *
+     * @param line Line to get an intersection from
+     * @return std::shared_ptr<Vector2> Vector representation of this intersection
+     */
+    [[nodiscard]] std::optional<Vector2> intersects(const Line &line) const;
+
+    /**
+     * @brief Gets the intersection of the lines
+     * See https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection for help
+     *
+     * @param line LineSegment to get an intersection from
+     * @return std::shared_ptr<Vector2> Vector representation of this intersection
+     */
+    [[nodiscard]] std::optional<Vector2> intersects(const LineSegment &line) const;
+
+    /**
+     * @brief Compute the line intersection of this line and the given as parameter line in positive direction.
+     * Which means in the direction from the start to the end of the given parameter line. The intersection position does not have to be in positive direction compared to this
+     * line, only to the given parameter line. Note if the line given as parameter is equal to this line then no intersection is returned.
+     *
+     * @param line The line of which the positive direction intersection is computed.
+     * @return No vector if the lines do not intersect in positive direction. Otherwise it returns a vector which represents the position of intersection.
+     */
+    [[nodiscard]] std::optional<Vector2> forwardIntersect(const Line &line) const;
+
+    /**
+     * @brief Checks whether \ref line intersects `this`
+     *
+     * @param line Line to check against
+     * @return true if \ref line intersects `this`
+     * @return false False if \ref line does not intersect `this`
+     */
+    [[nodiscard]] bool doesIntersect(const Line &line) const;
+
+    /**
+     * @brief Checks whether \ref line intersects `this`
+     *
+     * @param line Line to check against
+     * @return true True if \ref line intersects `this`
+     * @return false False if \ref line does not intersect `this`
+     */
+    [[nodiscard]] bool doesIntersect(const LineSegment &line) const;
+
+   private:
+    /**
+     * Get the intersection point between this line and the given line as parameter. Moreover compute how much the given line is extended relatively in the direction of
+     * the start of the line towards the end of the line to find the intersection point. Note if the line given as parameter is equal to this line then no intersection is
+     * returned.
+     *
+     * @param line Line to check against.
+     * @return std::nullopt if the lines do not intersect. Otherwise returns the intersection point and the relative extension of the given line to reach the intersection
+     * point.
+     */
+    [[nodiscard]] std::optional<std::pair<Vector2, float>> generalIntersect(const Line &line) const;
```

Comment:
Code clones are very dangerous...

Reply:
I just disagree with your statement here. This behavior has been tested and verified before. Now you are trading performance on critical functions, coupling code, and potentially breaking things all for the sake of readability.

Reply:
Well not tested enough when certainly bugs have shown up. Also new developers don't know whether a function has been verified and to which degree, so when we look at these functions we will try to verify them again.

> Now you are trading performance on critical functions, coupling code, and potentially breaking things all for the sake of readability.
I am trading a tiny bit performance to get more readable code, to prevent future bugs, to be consistent, to have a better design, etc. Code clones are far more dangerous than having a little bit more time overhead. Teacher in the past always told me first focus on your design then on performance. And the last step is often not needed, because if you have good design then in most cases you automatically also have a good performance.

Reply:
And what do you mean by "extending a line"? Isn't a line infinitely long?